core/src/spotify_id.rs

```diff
@@ -35,29 +46,59 @@ impl SpotifyId {
         let mut n: u128 = u128::zero();
         for c in data {
             let d = match BASE62_DIGITS.iter().position(|e| e == c) {
-                None => return Err(SpotifyIdError),
+                None => {
+                    warn!("Got an error decoding");
+                    return Err(SpotifyIdError)
+                },
                 Some(x) => x as u64,
             };
             n = n * u128::new(62);
             n = n + u128::new(d);
         }
-
-        Ok(SpotifyId(n))
+        Ok(SpotifyId::Track(n))
     }
 
+    // Shouldn't be used, from_rawURI should be the only usage
     pub fn from_raw(data: &[u8]) -> Result<SpotifyId, SpotifyIdError> {
+        debug!("Called from_raw()");
         if data.len() != 16 {
             return Err(SpotifyIdError);
         };
 
         let high = BigEndian::read_u64(&data[0..8]);
         let low = BigEndian::read_u64(&data[8..16]);
+        Ok(SpotifyId::Track(u128::from_parts(high, low)))
+    }
 
-        Ok(SpotifyId(u128::from_parts(high, low)))
+    // for episode support
+    // extract the spotify id from the uri
+    pub fn from_rawURI(data: &str) -> Result<SpotifyId, SpotifyIdError> {
+        let parts = data.split(":");
+        let vec = parts.collect::<Vec<&str>>();
+        let uri = vec.last().unwrap();
+        let spotify_type = vec[1];
+        let id = SpotifyId::from_base62(uri);
+        if spotify_type == "episode" {
+            let n = match id.unwrap() {
+                SpotifyId::Track(n) => n,
+                SpotifyId::NonPlayable(n) => n,
+                SpotifyId::Episode(n) => n,
+            };
+            Ok(SpotifyId::Episode(n))
+        }
+        else {
+            id
+        }
     }
 
     pub fn to_base16(&self) -> String {
-        let &SpotifyId(ref n) = self;
+        // This code seems idiotic, but should do the job
+        // Need someone with rust expertise to do better
```

Comment:
@peterjoel

Reply:
If every variant has a common field then perhaps this form of enum could be the wrong choice. You could consider something like this:

```rust
pub enum SpotifyIdType {
    Track,
    Episode,
    NonPlayable,
}

pub struct SpotifyId {
    value: u128,
    id_type: SpotifyIdType,
}
```

But what you have is not _so_ strange or uncommon in Rust code. Keeping the data structure as it is, it would be "better" to put the match expression in a method of `SpotifyId`, so it can be reused. For example: https://stackoverflow.com/a/54866005/493729